### PR TITLE
DOC: Adds examples for dtype attributes

### LINKED ISSUE #13702: DOC : dype attributes docs don't contain examples
The documentation for attributes of `dtype` in the `_add_newdocs.py` can be improved by adding examples, which are currently missing for most of the attributes.

### PR DESCRIPTION
**Pull request by kritisingh1**
Fixes #13702 
I fixed the docstrings for `numpy.dtype.base` in this PR as well.

### CODE REVIEW

File: numpy/core/_add_newdocs.py [hunk 1]
```diff
@@ -5482,6 +5514,13 @@
     of these flags is in C-API documentation; they are largely useful
     for user-defined data-types.
 
+    Examples
+    --------
+
+    >>> x = np.dtype([('a', np.int32, 8), ('b', np.float64, 6)])
+    >>> x.flags
+    16
```
Comment (eric-wieser) [inline numpy/core/_add_newdocs.py]: Worth showing that this is equal to `np.core.multiarray.NEEDS_PYAPI`, perhaps by adding on the line below:
```
>>> np.core.multiarray.NEEDS_PYAPI
16
```

Comment (kritisingh1) [reply numpy/core/_add_newdocs.py]: @eric-wieser I think we should mention the correlation between the two if we are going to mention this. And I am not sure I understand that. Could you clarify this a little bit more?

File: numpy/core/_add_newdocs.py [hunk 2]
```diff
@@ -5599,13 +5675,31 @@
...
 
 add_newdoc('numpy.core.multiarray', 'dtype', ('shape',
     """
     Shape tuple of the sub-array if this data type describes a sub-array,
     and ``()`` otherwise.
 
+    Examples
+    --------
+
+    >>> dt = np.dtype('4i4')
```
Comment (eric-wieser) [inline numpy/core/_add_newdocs.py]: I'd prefer to use the more verbose `('i4', 4)` syntax here

File: numpy/core/_add_newdocs.py (hunk 1 above)
Comment (eric-wieser) [reply numpy/core/_add_newdocs.py]: We already mention that correlation a few lines above

File: numpy/core/_add_newdocs.py (hunk 2 above)
Comment (kritisingh1) [reply numpy/core/_add_newdocs.py]: Doing this results in shape `()` . Should I include that example?

Comment (mattip) [reply numpy/core/_add_newdocs.py]: I think he meant `dt = np.dtype(('i4', 4))`

File: numpy/core/_add_newdocs.py [hunk 3]
```diff
@@ -5615,6 +5714,17 @@
 
     .. versionadded:: 1.13.0
 
+    Examples
+    ---------
+
+    >>> x = np.dtype('8f')
```
Comment (eric-wieser) [inline numpy/core/_add_newdocs.py]: `np.dtype((float, 8))` I think would be clearer. Let's not require the user to understand the multitude of different string mini-languages to grok these examples..

Can you add another test of `np.dtype((float, (3, 4)))` or similar, which constructs a 2d subarray dtype?

File: numpy/core/_add_newdocs.py [hunk 4]
```diff
@@ -5599,13 +5680,31 @@
...
     """
     Shape tuple of the sub-array if this data type describes a sub-array,
     and ``()`` otherwise.
 
+    Examples
+    --------
+
+    >>> dt = np.dtype('4i4')
+    >>> dt.shape
+    (4,)
```
Comment (eric-wieser) [inline numpy/core/_add_newdocs.py]: Can you also add `np.dtype(('i4', (2, 3)))` here, showing what more than 1d looks like?

**Comment by mattip** [comment]
Thanks @kritisingh1

**Pull request merged**